src/lib/FocusTrap.ts

```diff
@@ -0,0 +1,137 @@
+
+    lastFocusableElement: HTMLElement | null = null;
+
+    trapFocusableElement: HTMLElement | null = null;
+
+    constructor(element: HTMLElement) {
+        this.element = element;
+    }
+
+    destroy(): void {
+        this.disable();
+    }
+
+    focusFirstElement = (): void => {
+        if (!this.element) {
+            return;
+        }
+
+        const focusableElements = this.getFocusableElements();
+        if (focusableElements.length > 0) {
+            focusableElements[0].focus();
+        } else if (this.trapFocusableElement) {
+            this.trapFocusableElement.focus();
+        }
+    };
+
+    focusLastElement = (): void => {
+        if (!this.element) {
+            return;
+        }
+
+        const focusableElements = this.getFocusableElements();
+        if (focusableElements.length > 0) {
+            focusableElements[focusableElements.length - 1].focus();
+        } else if (this.trapFocusableElement) {
+            this.trapFocusableElement.focus();
+        }
+    };
+
+    getFocusableElements = (): Array<HTMLElement> => {
+        // Look for focusable elements
+        const foundElements = this.element.querySelectorAll(FOCUSABLE_ELEMENTS.join(', '));
+        // Filter out the elements that are preview control buttons and not visible
+        return Array.prototype.slice.call(foundElements).filter(el => (isButton(el) && isVisible(el)) || !isButton(el));
+    };
+
+    handleTrapKeydown = (event: KeyboardEvent): void => {
+        const key = decodeKeydown(event);
+        const isTabPressed = key === 'Tab' || key === 'Shift+Tab';
+
+        if (!isTabPressed) {
+            return;
+        }
+        event.stopPropagation();
+        event.preventDefault();
+    };
+
+    handleKeydown = (event: KeyboardEvent): void => {
+        const key = decodeKeydown(event);
+        const isTabPressed = key === 'Tab' || key === 'Shift+Tab';
+
+        if (this.element === document.activeElement && isTabPressed) {
+            this.focusFirstElement();
+            event.stopPropagation();
+            event.preventDefault();
+        }
+    };
+
+    enable(): void {
+        this.element.addEventListener('keydown', this.handleKeydown);
+
+        // Create focus anchors (beginning, end and trap)
+        this.firstFocusableElement = createFocusAnchor();
+        this.lastFocusableElement = createFocusAnchor();
+        this.trapFocusableElement = createFocusAnchor();
+
+        this.firstFocusableElement.addEventListener('focus', this.focusLastElement);
+        this.lastFocusableElement.addEventListener('focus', this.focusFirstElement);
+        this.trapFocusableElement.addEventListener('keydown', this.handleTrapKeydown);
+
+        this.element.prepend(this.firstFocusableElement);
+        this.element.append(this.lastFocusableElement);
+        this.element.append(this.trapFocusableElement);
+    }
+
+    disable(): void {
+        this.element.removeEventListener('keydown', this.handleKeydown);
```

Comment:
We have guard clauses for `if (!this.element)` in some of our other methods. Do we need them everywhere?

Reply:
I'll simplify it by throwing an error in the constructor if we don't get an element so I can remove all guard clauses